<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-del-projecto-equipo-52/blob/main/Semana%203%20-%20Actividad%201.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

> # **Semana 3 - Actividad 1**
>## **Maestría en Inteligencia Artificial Aplicada**
>### **Curso: TC4029.10 Ciencia y analítica de datos (Gpo 10)**
>## **Tecnológico de Monterrey**
>## **Prof: Jobish Vallikavungal Devassia**




### **Gerardo Aaron Castañeda Jaramillo  -  Matrícula A01137646**
### **Genaro Ramos Higuera  -  Matrícula A00351269**
### **30 de septiembre del 2022**

### **Parte 1: Fundamentos de bases de datos**

**1. Fundamentos de bases de datos y para ciencia de datos.**

Una base de datos es un conjunto de datos estructurados y almacenado ya sea localmente en computadoras o servidores o en la nube (Cloud) y permite ser accesado de diferentes formas para poder actualizar la información necesaria para la empresa, institución u organización. 

Existen muchas aplicaciones que permiten organizar datos en forma estructurada. Las bases de datos relacionales (*SQL Databases*) son aquellas que almacenan y proporcionan acceso a puntos de datos en tablas relacionados entre sí.

Entre las más populares SQL son : Oracle, MySQL, MS SQL Server, PostgreSQL, Amazon Redshift, IBM DB2 ,MS Access, SQLite y Snowflake.

 El modelo relacional ayuda a mantener uniformidad en los datos y en todas las aplicaciones y sus instancias. Mientras que las bases de datos no relacionales, también conocidas como NoSQL databases, son las que conectan la información guardada en ellas por categorías, en vez de relaciones.

Entre las más populares NoSQL están: MongoDB, Apache Cassandra, Apache HBase
Apache CouchDB, Neo4j, RavenDB, Redis, OrientDB, DynamoDB y HyperTable.

  *SQL (Structured Query Language*) es un lenguaje de programación utilizado en las bases de datos relacionales. También es muy utilizado en la ciencia de datos. Los datos son la parte mas importante en la ciencia de datos, por lo que manejar bases de datos es esencial para la misma. No se puede tener Ciencia de Datos sin Datos.

Entre los lenguajes de programación más utilizados en la Ciencia y Analítica de Datos están: **Python,R, SQL, Java, Scala, Julia y Matlab**.




**2. Fundamentos de almacenes de datos (Data Warehouse) para ciencia de datos.**

Un "almacén de datos”, mejor conocido como data warehouse, es una base de datos gigante que junta datos de múltiples y diferentes fuentes. Es un sistema diseñado para permitir y soportar las actividades de Ciencia y Analítica de Datos, ya que contiene una gran cantidad de datos históricos. Uno de sus retos es el tomar los datos de las diversas fuentes, transformarlos dentro de un esquema que permita consultarlos simultáneamente. 

Así como la normalización de los datos. Es necesario relacionar de alguna manera los datos de múltiples fuentes entre ellos para poder trabajarlos. Dentro de data warehousing existen dos términos importantes: ETL (extract, transform and load) y ELT (extract, load and transform). ETL consiste como sugiere su nombre, es extraer los datos, transformarlos y después cargarlos a la data warehouse. La cuestión es que, si los datos alcanzan magnitudes muy grandes, el transformarlos puede ser un cuello de botella. Aquí entra ELT, que los extrae, los carga, y ya cargados los transforma. Es algo muy útil para la ciencia de datos, ya que es posible utilizar el procesamiento del repositorio para la transformación de datos. 


### **Parte 2: Selección y limpieza de los Datos en Python**

Comenzamos por importar las librerías para trabajar

In [200]:
import pandas as pd
import numpy as np

Proseguimos convirtiendo el archivo csv a trabajar a una data frame

In [201]:
df = pd.read_csv('https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv')
df

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

Primero creo un diccionario para dejar en claro que significa cada columna del dataframe. En base a que es cada columna se tomaran decisiones sobre que técnicas de manejo para datos faltantes se usaran

Esta es la informacion del Dataframe que tiene 30,000 registros, y notamos que algunos tienen registros con algunas columnas con datos nulos:

In [202]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      29999 non-null  float64
 3   X3      29998 non-null  float64
 4   X4      29998 non-null  float64
 5   X5      29995 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29995 non-null  float64
 8   X8      29993 non-null  float64
 9   X9      29991 non-null  float64
 10  X10     29984 non-null  float64
 11  X11     29986 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

In [204]:
definition = ['ID',
              'Amount of the given credit (NT dollar): it includes both the individual consumer credit and his/her family (supplementary) credit. ',
              'Gender (1 = male; 2 = female)',
              'Education (1 = graduate school; 2 = university; 3 = high school; 4 = others)',
              'Marital status (1 = married; 2 = single; 3 = others)',
              'Age (year)',
              'the repayment status in September, 2005, The measurement scale for the repayment status is: -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above',
              'the repayment status in August, 2005',
              'the repayment status in July, 2005',
              'the repayment status in June, 2005',
              'the repayment status in May, 2005',
              'the repayment status in April, 2005',
              'amount of bill statement in September, 2005',
              'amount of bill statement in August, 2005',
              'amount of bill statement in July, 2005',
              'amount of bill statement in June, 2005',
              'amount of bill statement in May, 2005',
              'amount of bill statement in April, 2005',
              'amount paid in September, 2005',
              'amount paid in August, 2005',
              'amount paid in July, 2005',
              'amount paid in June, 2005',
              'amount paid in May, 2005',
              'amount paid in April, 2005',
              'the real probability of default as the response variable']
meanings = {}
n = 0
for col in df.columns:
  meanings[col] = definition[n]
  n = n + 1
meanings


{'ID': 'ID',
 'X1': 'Amount of the given credit (NT dollar): it includes both the individual consumer credit and his/her family (supplementary) credit. ',
 'X2': 'Gender (1 = male; 2 = female)',
 'X3': 'Education (1 = graduate school; 2 = university; 3 = high school; 4 = others)',
 'X4': 'Marital status (1 = married; 2 = single; 3 = others)',
 'X5': 'Age (year)',
 'X6': 'the repayment status in September, 2005, The measurement scale for the repayment status is: -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above',
 'X7': 'the repayment status in August, 2005',
 'X8': 'the repayment status in July, 2005',
 'X9': 'the repayment status in June, 2005',
 'X10': 'the repayment status in May, 2005',
 'X11': 'the repayment status in April, 2005',
 'X12': 'amount of bill statement in September, 2005',
 'X13': 'amount of bill statement in August, 2005',
 'X14': 'amount of bill sta

Primeramente verificamos que en nuestro archivo no se encuentren ID repetidos, y esto lo hacemos con las siguientes instrucciones:

In [205]:
if df['ID'].count() == df["ID"].unique().size:
  print ("Tenemos los siguientes registros ID únicos: ", df["ID"].unique().size)
else: 
  print ("Existen registros algunos repetidos y solo tenemos registros ID únicos: ", df["ID"].unique().size)

Tenemos los siguientes registros ID únicos:  30000


Seguido, vamos a trabajar con valores de Monto de Crédito Otorgados que sean válidos. Comenzamos verificando que X1 incluyan solo montos positivos y lo verificamos de la manera siguiente:

In [206]:
df_norm = df.copy()
(df_norm['X1'].values < 0).sum()

0

O también podemos hacerlo de esta otra manera:

In [207]:
df_norm.loc[df['X1'] < 0] # para saber que registros tienen un monto negativo

Empty DataFrame
Columns: [ID, X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, Y]
Index: []

[0 rows x 25 columns]

In [208]:
df_norm.head(10
             )

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...       X15  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...       0.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...    3272.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...   14331.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...   28314.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...   20940.0   
5   6   50000  1.0  1.0  2.0  37.0  0.0  0.0  0.0  0.0  ...   19394.0   
6   7  500000  1.0  1.0  2.0  29.0  0.0  0.0  0.0  0.0  ...  542653.0   
7   8  100000  2.0  2.0  2.0  23.0  0.0 -1.0 -1.0  0.0  ...     221.0   
8   9  140000  2.0  3.0  1.0  28.0  0.0  0.0  2.0  0.0  ...   12211.0   
9  10   20000  1.0  3.0  2.0  35.0 -2.0 -2.0 -2.0 -2.0  ...       0.0   

        X16       X17      X18      X19      X20      X21      X22      X23  \
0       0.0       0.0      0.0    689.0      0.0      0.0      0.0      0.0   
1    3455.0    3261.0      0.0   1000.0   1000.0   1000.0      0.0   2000.0   
2   14948.0   15549.0   1518.0   1500.0   1000.0   1000.0   1000.0   5000.0   
3   28959.0   29547.0   2000.0   2019.0   1200.0   1100.0   1069.0   1000.0   
4   19146.0   19131.0   2000.0  36681.0  10000.0   9000.0    689.0    679.0   
5   19619.0   20024.0   2500.0   1815.0    657.0   1000.0   1000.0    800.0   
6  483003.0  473944.0  55000.0  40000.0  38000.0  20239.0  13750.0  13770.0   
7    -159.0     567.0    380.0    601.0      0.0    581.0   1687.0   1542.0   
8   11793.0    3719.0   3329.0      0.0    432.0   1000.0   1000.0   1000.0   
9   13007.0   13912.0      0.0      0.0      0.0  13007.0   1122.0      0.0   

     Y  
0  1.0  
1  1.0  
2  0.0  
3  0.0  
4  0.0  
5  0.0  
6  0.0  
7  0.0  
8  0.0  
9  0.0  

[10 rows x 25 columns]

Seguido, checamos que X2 solo sean valores de 1 o 2

In [209]:
(df_norm['X2'].values < 1).sum() + (df_norm['X2'].values > 2).sum()

0

In [210]:
hombres=df_norm.loc[df_norm['X2'] == 1,"X2"].count()
mujeres=df_norm.loc[df_norm['X2'] == 2,"X2"].count()
prop_hombres=hombres/(hombres+mujeres)
prop_mujeres=mujeres/(hombres+mujeres)

print (" De los cuales hombres son ",hombres," y mujeres ", mujeres)
print ("Con proporcion de hombres de ", prop_hombres, "y de mujeres de ", prop_mujeres, " respectivamente.")

 De los cuales hombres son  11887  y mujeres  18112
Con proporcion de hombres de  0.39624654155138506 y de mujeres de  0.6037534584486149  respectivamente.


Sabemos que hay 30,000 registros esto significa que uno de ellos es Nulo

In [211]:
# creamos una serie booleana para encontrar el registro nulo
sexo_nulo = pd.isnull(df_norm["X2"])
   
# filtramos los datos
# desplegamos el datos con "X2" = NaN
df_norm[sexo_nulo]


ID      X1  X2  X3  X4  X5  X6  X7  X8  X9  ...  X15  X16  X17  X18  \
24365  24366  130000 NaN NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  NaN  NaN   

       X19  X20  X21  X22    X23    Y  
24365  NaN  NaN  NaN  NaN  390.0  0.0  

[1 rows x 25 columns]

In [212]:
#Ahora reeemplazamos el registro con sexo nulo dandole el 2 por ser el mayor porcentaje con el 60%
df_norm["X2"].fillna(2,inplace=True)
df_norm.loc[24365:24365]

ID      X1   X2  X3  X4  X5  X6  X7  X8  X9  ...  X15  X16  X17  \
24365  24366  130000  2.0 NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  NaN   

       X18  X19  X20  X21  X22    X23    Y  
24365  NaN  NaN  NaN  NaN  NaN  390.0  0.0  

[1 rows x 25 columns]

Ahora seguimos con X3, que son el nivel de educación solo debe admitir valores de 1, 2, 3, y 4 para cualquier otro. Así que verificamos

In [213]:
print('Valores menores a 1: ' + str((df_norm['X3'].values < 1).sum()))
print('Valores mayores a 4: ' + str((df_norm['X3'].values > 4).sum()))

Valores menores a 1: 14
Valores mayores a 4: 331


Vemos que hay 331 valores mayores a 4. En este caso tomaremos la suposición de que todo valor mayor a 4 puede entrar dentro del valor 4, que significa "4 = others".

In [214]:
X3_values = []
for i in range(0,len(df_norm.index)):
  if df_norm['X3'].iloc[i] > 4:
    if df_norm['X3'].iloc[i] not in X3_values:
      X3_values.append(int(df_norm['X3'].iloc[i]))
for i in X3_values:
  df_norm['X3'].replace(i,4,inplace=True)

print('Valores mayores a 4: ' + str((df_norm['X3'].values > 4).sum()))

Valores mayores a 4: 0


Mientras que los valores menores a 1, que solo son 14, los remplazaremos el promedio

In [215]:
X3_mean = df_norm.X3.mean()
print('Promedio de X3: ' + str(int(X3_mean)))
df_norm['X3'].replace(0,int(X3_mean),inplace=True)
print('Valores menores a 1: ' + str((df_norm['X3'].values < 1).sum()))

Promedio de X3: 1
Valores menores a 1: 0


In [216]:
# creamos una serie booleana para encontrar registros nulos en X3
educacion_nulos = pd.isnull(df_norm["X3"])
   
# filtramos los datos
# desplegamos los registros con "X3" = NaN
df_norm[educacion_nulos]

ID      X1   X2  X3  X4  X5  X6  X7  X8  X9  ...  X15  X16      X17  \
6227    6228   30000  1.0 NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  21120.0   
24365  24366  130000  2.0 NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN      NaN   

          X18     X19     X20     X21     X22    X23    Y  
6227   1586.0  1365.0  1663.0  5106.0  1050.0    0.0  0.0  
24365     NaN     NaN     NaN     NaN     NaN  390.0  0.0  

[2 rows x 25 columns]

In [217]:
# les asignamos la media de educacion = 1 a los 2 registros nulos encontrados
df_norm["X3"].fillna(1,inplace=True)

In [218]:
df_norm.loc[24365:24365]


ID      X1   X2   X3  X4  X5  X6  X7  X8  X9  ...  X15  X16  X17  \
24365  24366  130000  2.0  1.0 NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  NaN   

       X18  X19  X20  X21  X22    X23    Y  
24365  NaN  NaN  NaN  NaN  NaN  390.0  0.0  

[1 rows x 25 columns]

In [219]:
df_norm.loc[6227:6227]

ID     X1   X2   X3  X4  X5  X6  X7  X8  X9  ...  X15  X16      X17  \
6227  6228  30000  1.0  1.0 NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  21120.0   

         X18     X19     X20     X21     X22  X23    Y  
6227  1586.0  1365.0  1663.0  5106.0  1050.0  0.0  0.0  

[1 rows x 25 columns]

No debiesen haber registros nulos en Educacion X3 y lo comprobamos así:

In [220]:
educacion_nulos = pd.isnull(df_norm["X3"])
   
# filtramos los datos
# desplegamos los registros con "X3" = NaN
df_norm[educacion_nulos] 

Empty DataFrame
Columns: [ID, X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, Y]
Index: []

[0 rows x 25 columns]

X4 solo debería admitir valores de 1, 2, y 3. así que verificamos

In [221]:
print('Valores menores a 1: ' + str((df_norm['X4'].values < 1).sum()))
print('Valores mayores a 3: ' + str((df_norm['X4'].values > 3).sum()))

Valores menores a 1: 54
Valores mayores a 3: 0


Descubrimos 54 valores de 0, así que sustituimos por el promedio

In [222]:
X4_mean = df_norm.X4.mean()
print('Promedio de X4: ' + str(int(X4_mean)))
df_norm['X4'].replace(0,int(X4_mean),inplace=True)
print('Valores menores a 1: ' + str((df_norm['X4'].values < 1).sum()))

Promedio de X4: 1
Valores menores a 1: 0


In [223]:
# creamos una serie booleana para encontrar registros nulos en X4
marital_status_nulos = pd.isnull(df_norm["X4"])
   
# filtramos los datos
# desplegamos los registros con "X4" = NaN
df_norm[marital_status_nulos]

ID      X1   X2   X3  X4  X5  X6  X7  X8  X9  ...  X15  X16  \
6227    6228   30000  1.0  1.0 NaN NaN NaN NaN NaN NaN  ...  NaN  NaN   
24365  24366  130000  2.0  1.0 NaN NaN NaN NaN NaN NaN  ...  NaN  NaN   

           X17     X18     X19     X20     X21     X22    X23    Y  
6227   21120.0  1586.0  1365.0  1663.0  5106.0  1050.0    0.0  0.0  
24365      NaN     NaN     NaN     NaN     NaN     NaN  390.0  0.0  

[2 rows x 25 columns]

In [224]:
# les asignamos la media de Marital Status = 1 Married a los 2 registros nulos encontrados
df_norm["X4"].fillna(1,inplace=True)

In [225]:
df_norm.loc[6227:6227, "ID":"X4"]

ID     X1   X2   X3   X4
6227  6228  30000  1.0  1.0  1.0

In [226]:
df_norm.loc[24365:24365, "ID":"X4"]

ID      X1   X2   X3   X4
24365  24366  130000  2.0  1.0  1.0

Ahora checamos X5. Donde no debería haber menores a 18

In [227]:
print('Valores menores a 18: ' + str((df_norm['X5'].values < 18).sum()))

Valores menores a 18: 0


In [228]:
# creamos una serie booleana para encontrar registros nulos en X5
edad_nulos = pd.isnull(df_norm["X5"])
   
# filtramos los datos
# desplegamos los registros con "X5" = NaN
df_norm[edad_nulos]

ID      X1   X2   X3   X4  X5   X6   X7   X8   X9  ...       X15  \
182      183  500000  2.0  1.0  1.0 NaN  0.0  0.0  0.0  0.0  ...  122967.0   
219      220  310000  2.0  1.0  2.0 NaN -1.0 -1.0 -1.0 -1.0  ...       0.0   
239      240  140000  2.0  2.0  3.0 NaN  0.0  0.0  0.0  0.0  ...   19068.0   
6227    6228   30000  1.0  1.0  1.0 NaN  NaN  NaN  NaN  NaN  ...       NaN   
24365  24366  130000  2.0  1.0  1.0 NaN  NaN  NaN  NaN  NaN  ...       NaN   

            X16      X17      X18      X19      X20      X21      X22  \
182    108834.0  70064.0  70010.0  30357.0  30000.0  20000.0  52183.0   
219         0.0      0.0   4542.0    126.0      0.0      0.0      0.0   
239     16409.0  16383.0   3000.0   2000.0   2198.0   1000.0   3000.0   
6227        NaN  21120.0   1586.0   1365.0   1663.0   5106.0   1050.0   
24365       NaN      NaN      NaN      NaN      NaN      NaN      NaN   

           X23    Y  
182    20000.0  0.0  
219        0.0  0.0  
239     2399.0  1.0  
6227       0.0  0.0  
24365    390.0  0.0  

[5 rows x 25 columns]

In [229]:
promedio_edad=int(df_norm['X5'].mean())
promedio_edad

35

Reemplazamos los 5 registros nulos por la edad promedio de 35 años, ya que apenas son el 0.016% del total de registros.

In [230]:
df_norm["X5"].fillna(promedio_edad,inplace=True) 

In [231]:
df_norm.loc[182:182, "ID":"X5"]

ID      X1   X2   X3   X4    X5
182  183  500000  2.0  1.0  1.0  35.0

In [232]:
df_norm.loc[219:219, "ID":"X5"]

ID      X1   X2   X3   X4    X5
219  220  310000  2.0  1.0  2.0  35.0

In [233]:
df_norm.loc[239:239, "ID":"X5"]

ID      X1   X2   X3   X4    X5
239  240  140000  2.0  2.0  3.0  35.0

In [234]:
df_norm.loc[6227:6227, "ID":"X5"]

ID     X1   X2   X3   X4    X5
6227  6228  30000  1.0  1.0  1.0  35.0

In [235]:
df_norm.loc[24365:24365, "ID":"X5"]

ID      X1   X2   X3   X4    X5
24365  24366  130000  2.0  1.0  1.0  35.0

Para X6-X11, no deben haber valores menores a -1, ni mayores a 9

In [236]:
print('Valores menores a -1: ' + str((df_norm.iloc[:,6:12].values < -1).sum()))
print('Valores mayores a 9: ' + str((df_norm.iloc[:,6:12].values > 9).sum()))

Valores menores a -1: 24415
Valores mayores a 9: 0


In [237]:
print('Encontramos que son iguales a -2: ' + str((df_norm.iloc[:,6:12].values == -2).sum()))

Encontramos que son iguales a -2: 24415


En este caso, asumiremos que los Valores Menores a -1, esto es -2, equivalen a -1 y se asume o que son errores de captura, o significan que llevan periodos consecutivos cumpliendo. Ya que -1 dice que se pagó a tiempo, y cualquier valor mayor a ese son los meses que se han atrasado en pagos.

In [263]:
df_norm.iloc[:,6:12] = df_norm.iloc[:,6:12].replace(-2,-1)
print('Valores menores a -1: ' + str((df_norm.iloc[:,6:12].values < -1).sum()))

Valores menores a -1: 0


In [297]:
(df_norm.iloc[:,6:7].values == 0).sum()

14736

Ahora veamos a los registros con valores NULOS en las Columnas X6 a X11


In [300]:
keys=["X6","X7","X8","X9","X10","X11"]


for i in keys:
  nulos = pd.isnull(df_norm[i])
  print(df_norm[nulos])
 
  

          ID      X1   X2   X3   X4    X5  X6  X7  X8  X9  ...  X15  X16  \
6227    6228   30000  1.0  1.0  1.0  35.0 NaN NaN NaN NaN  ...  NaN  NaN   
24123  24124   50000  1.0  1.0  2.0  29.0 NaN NaN NaN NaN  ...  NaN  NaN   
24365  24366  130000  2.0  1.0  1.0  35.0 NaN NaN NaN NaN  ...  NaN  NaN   

           X17     X18     X19     X20     X21     X22    X23    Y  
6227   21120.0  1586.0  1365.0  1663.0  5106.0  1050.0    0.0  0.0  
24123      NaN     NaN     NaN     NaN     NaN     NaN  360.0  0.0  
24365      NaN     NaN     NaN     NaN     NaN     NaN  390.0  0.0  

[3 rows x 25 columns]
          ID      X1   X2   X3   X4    X5   X6  X7  X8  X9  ...  X15  X16  \
6227    6228   30000  1.0  1.0  1.0  35.0  NaN NaN NaN NaN  ...  NaN  NaN   
24123  24124   50000  1.0  1.0  2.0  29.0  NaN NaN NaN NaN  ...  NaN  NaN   
24365  24366  130000  2.0  1.0  1.0  35.0  NaN NaN NaN NaN  ...  NaN  NaN   
29832  29833   20000  1.0  2.0  2.0  30.0  0.0 NaN NaN NaN  ...  NaN  NaN   
29833  2983

In [280]:
df_norm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      30000 non-null  float64
 3   X3      30000 non-null  float64
 4   X4      30000 non-null  float64
 5   X5      30000 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29995 non-null  float64
 8   X8      29993 non-null  float64
 9   X9      29991 non-null  float64
 10  X10     29984 non-null  float64
 11  X11     29986 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

Para X12-X23 checamos que no existan valores negativos

In [194]:
print('Valores menores a 0: ' + str((df_norm.iloc[:,12:24].values < 0).sum()))

Valores menores a 0: 3932


En este caso, asumiremos que es un error de captura y son valores positivos. En la realidad tendríamos que verificar como se capturaron los datos para asegurarnos que sea posible hacer esto, o entender porque existen valores negativos y qué hacer con ellos

In [195]:
for col in df_norm.iloc[:,12:24].columns:
  df_norm[col] = df_norm[col].abs()
print('Valores menores a 0: ' + str((df_norm.iloc[:,12:24].values < 0).sum()))

Valores menores a 0: 0


Y finalmente, vemos una descripción del dataframe ya trabajado:

In [196]:
df_norm.describe()

ID              X1            X2            X3            X4  \
count  30000.000000    30000.000000  30000.000000  30000.000000  30000.000000   
mean   15000.500000   167484.322667      1.603767      1.840733      1.553667   
std     8660.398374   129747.661567      0.489122      0.743227      0.518326   
min        1.000000    10000.000000      1.000000      1.000000      1.000000   
25%     7500.750000    50000.000000      1.000000      1.000000      1.000000   
50%    15000.500000   140000.000000      2.000000      2.000000      2.000000   
75%    22500.250000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      4.000000      3.000000   

                 X5            X6            X7            X8            X9  \
count  30000.000000  29997.000000  29995.000000  29993.000000  29991.000000   
mean      35.484133      0.075341     -0.007601     -0.030207     -0.075823   
std        9.217258      0.990785      1.035836      1.024061      0.987410   
min       21.000000     -1.000000     -1.000000     -1.000000     -1.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

       ...            X15            X16            X17            X18  \
count  ...   29985.000000   29983.000000   29990.000000   29992.000000   
mean   ...   43349.094981   40395.916719   39022.199500    5662.945886   
std    ...   64296.043549   60762.560911   59468.986025   16564.165089   
min    ...       0.000000       0.000000       0.000000       0.000000   
25%    ...    2400.000000    1867.500000    1359.250000    1000.000000   
50%    ...   19068.000000   18128.000000   17135.000000    2100.000000   
75%    ...   54589.000000   50231.500000   49262.750000    5006.000000   
max    ...  891586.000000  927171.000000  961664.000000  873552.000000   

                X19            X20            X21            X22  \
count  2.999100e+04   29992.000000   29989.000000   29989.000000   
mean   5.922489e+03    5225.623400    4827.252526    4800.297209   
std    2.304418e+04   17608.422625   15668.751975   15280.842069   
min    0.000000e+00       0.000000       0.000000       0.000000   
25%    8.355000e+02     390.000000     296.000000     251.000000   
50%    2.009000e+03    1800.000000    1500.000000    1500.000000   
75%    5.000000e+03    4505.500000    4014.000000    4033.000000   
max    1.684259e+06  896040.000000  621000.000000  426529.000000   

                 X23             Y  
count   29995.000000  29997.000000  
mean     5216.259977      0.221189  
std     17778.848359      0.415054  
min         0.000000      0.000000  
25%       118.000000      0.000000  
50%      1500.000000      0.000000  
75%      4000.000000      0.000000  
max    528666.000000      1.000000  

[8 rows x 25 columns]

Ahora, verificamos si hay datos nulos

In [197]:
df_norm.isnull().values.any()

True

Vemos que columnas tienen datos faltantes

In [198]:
df_norm.isnull().any()

ID     False
X1     False
X2     False
X3     False
X4     False
X5     False
X6      True
X7      True
X8      True
X9      True
X10     True
X11     True
X12     True
X13     True
X14     True
X15     True
X16     True
X17     True
X18     True
X19     True
X20     True
X21     True
X22     True
X23     True
Y       True
dtype: bool

Analizamos cuantas filas tienen valores faltantes. Si el tamaño de la muestra es lo suficientemente grande, podemos aplicar eliminación por lista y asumir que tendremos estimaciones no sesgadas y resultados conservadores.

In [199]:
rows_missing = []
for i in range(0,len(df_norm.index)):
  if df_norm.iloc[i].isnull().any():
    rows_missing.append(i)
len(rows_missing)

KeyboardInterrupt: ignored

Y encontramos un total de 42 filas con algún dato faltante. Si consideramos que existe un total de 30000 filas, 42/30000 es un 0.14% de filas faltantes. Considero que podemos usar eliminación por lista.

In [ ]:
df_EL = df_norm.copy()
df_EL.dropna(inplace = True)

Verificamos que efectivamente no existan valores nulos

In [ ]:
df_EL.isna().values.any()

In [ ]:
df_EL

Y como resultado tenemos una nueva data frame de 29958 filas y 25 columnas donde no existe ningún valor nulo

In [ ]:
df.describe()

Tenemos una nueva data frame de 29958 filas y 25 columnas donde no existe ningún valor nulo
Ahora, con intención de aprender, analizaremos la dataframe original, pero con otros criterios. 
Suposiciones que tomaremos:
*   X1-X5 son parámetros del consumidor. Si faltan 2 o menos, podemos utilizar media o mediana para sustituir los valores faltantes. Si faltan 3 o más, se elimina la fila
*   X6-X11 es la historia de pagos. X12-17 es el bill statement en dólares. X18-X23 son los pagos previos en dólares. Y "Y" es la probabilidad del modelo. 
Dado que estos son datos faltantes no al azar y no vamos a modelar los datos que faltan, aprovecharemos que son pocos los que faltan y eliminaremos esas filas de los datos. 

Una muestra de los datos faltantes:

In [ ]:
df_rm = df_norm.iloc[rows_missing]
df_rm.head(10)

Primero eliminaremos todas las filas que tengan datos faltantes de X6 a Y

In [ ]:
df_EP = df_norm.copy()
df_EP.dropna(inplace = True,subset = df_EP.columns[6:26])
df_EP

Ahora vemos que columnas tienen datos faltantes:

In [ ]:
df_EP.isnull().any()

Encontramos que en este caso coinciden las filas con valores faltantes de X6-Y con los valores faltantes de muchas de las de X1-X4. Como resultado, solo X5 sigue teniendo valores nulos. X5 es la edad del cliente, un valor de tipo integer. Lo que podemos hacer aquí es sustituir por la media. Repito que mi primer criterio fue eliminación por lista, pero por utilizar otro método en el ejercicio haremos la sustitución por la media.

In [ ]:
media_edad = int(df_EP.X5.mean())
media_edad

Donde encontramos que es un valor 35

In [ ]:
df_EP['X5'].fillna(value = media_edad, inplace=True)
df_EP.isnull().any()

Con lo que nuestra dataframe ahora no tiene ningún valor nulo

In [ ]:
df_EP.describe()

Y donde podemos apreciar que a diferencia de las 29958 filas de la dataframe df_EL, donde se eliminaron todas las filas con algún valor nulo, contra el dataframe df_EP donde tenemos 29961 filas ya que 3 se sustituyeron sus valores nulos por la media

### **Parte 3: Preparación de los datos**
Con base en los resultados de tu libreta de Google Colab de la Parte 2 responde detalladamente las siguientes preguntas:

**1. ¿Qué datos considero mas importantes? ¿Por qué?**
> Todos los datos son importantes cuando buscamos general un modelo para el valor de salida en base a todos los valores de entrada. La mayoría de los algoritmos no aceptan valores faltantes. Por eso es necesario hacer limpieza de datos. No podemos tener una base de datos con valores nulos si la pensamos trabajar. En este caso, debido al tamaño de la muestra de los datos que tenemos vs. la cantidad de datos faltantes, fue posible aplicar la eliminación por lista. Lo que nos elimino 42 filas de datos. Fue valido ya que solo son el 0.14% del total de datos. Pero hay que evaluar métodos como la sustitución por la media. De ser posible aplicarse, nos deja una mayor cantidad de datos de entrada para el modelo. Siempre teniendo cuidado en no crear sesgos.  

**2. ¿Se eliminaron o reemplazaron datos nulos? ¿Qué se hizo y por qué?**
>Hice dos intentos diferentes. En el primero se verifico cuantas filas tenían algún dato faltante. Se encontró que eran 42. Considerando que el total son 30000 datos, 42 solo es el 0.14% de ellos. Por lo que considero que se pueden eliminar esas 42 filas con datos faltantes. Sin embargo, por cuestiones de práctica, hice otro intento diferente. En este segundo intento, se busco primero eliminar las filas de X6 a Y que tuvieran datos faltantes. Eliminando esas filas, se encontró que en X5, la edad, quedaron 3 filas que tenían ese dato faltante. Al considerar que la edad debe ser un valor tipo integer, se decidió hacer sustitución media. Donde se encontró que la media de la edad tenia un valor de 35, y se cambiaron los valores nulos por ese valor.  Considerando que la distribución de los datos sea normal, entonces es valido utilizar este método. 

**3. ¿Es necesario ordenar los datos para el análisis? Sí / No / ¿Por qué?**
> ¿Ordenarlos para el análisis de que métodos usar para eliminar o sustituir valores nulos? Si, es muy útil. Utilizando el método .describe() en el dataframe fue posible ver media, mínimos, máximos, desviación estándar y percentiles. Estos son valores muy útiles para analizar que esta pasando con los datos. Encontrar ruido, valores atípicos y valores inválidos nos da una mejor idea de como podemos tratar a los valores nulos. 

**4. ¿Existen problemas de formato que deban solucionar antes del proceso de modelado? Sí / No / Por qué.**
> Si, si existen. El valor de -2 en X5 no es válidos. Los valores posibles son -1, y de 1 a 9. -2 no nos dicen nada para el modelado. Así como los valores de X6 a X23. No pueden tener valores negativos, y si los hay. Algunas filas parecen no coincidir, cuando en teoría todas deben estar relacionadas. Si X6-X11 es la historia de los pagos hechos de los meses anteriores, X12-X17 es el estado de cuenta, y X18-23 la cantidad del pago hecho en su respectivo mes; entonces si el primer rango de Xs dice que se hicieron pagos, estos se deben de ver reflejados en el estado de cuenta, y debería salir de cuanto fue el pago hecho. Sin embargo, no coinciden estos valores en todas las filas. Si esto no se arregla, al momento de modelar se obtendrá un algoritmo no idóneo. 

**5. ¿Qué ajustes se realizaron en el proceso de limpieza de datos (agregar, integrar, eliminar, modificar registros (filas), cambiar atributos (columnas)?**
> En este caso comencé trabajando los valores atípicos. Cada columna tiene determinado que valores son posibles o no. Para X3, la columna de “Education”, se encontraron 14 valores menores a 1, y 331 mayores a 4. Se tomo la suposición de que todo valor mayor a 4 entra dentro del valor 4, ya que significa “4 = others”. Y se hizo la sustitución. Mientras que los valores menores a 1, que eran todos 0, se cambiaron por el promedio. En X3, la columna de estatus civil, solo existen las opciones para valores de 1, 2 y 3; pero se encontraron 54 valores de 0. Estos igual fueron sustituidos por el promedio. Para X6-11, no debe haber valores menores a -1. Por lo que se tomo la suposición de que todos los valores negativos equivalen a -1, ya que valor -1 equivale a que se hizo el pago a tiempo, y todo mayor 0 equivale a los meses que se ha atrasado el pago. En realidad, tendríamos que ver como se recolectaron los datos para confirmar esta suposición, pero para el ejercicio se hizo esa sustitución. Y para X12-X23 se cambiaron todos los valores negativos por su valor absoluto. 
Después de trabajar los valores, se hicieron 2 procesos diferentes. En el primero se eliminaron todas las filas que tuvieran datos faltantes, que equivalian a un 0.14% del total de datos. En el segundo caso se eliminaron las filas de X6 a X23 que tuvieran algún valor nulo, y luego se sustituyeron los valores nulos de X5 por la media de la edad.


**Referencias:**

Kane, F. (2017). Hands-On Data Science and Python Machine Learning. Packt Publising Ltd.

Teate, R. M. (2021). SQL for Data Scientiests. Wiley.

